File: Ruby/Basico/juego_ahorcado/ahorcado.rb
PALABRAS = %w[ruby programacion algoritmo variable metodo clase modulo iterador bloque simbolo]

DIBUJOS = [
  "  -----\n  |   |\n      |\n      |\n      |\n=========",
  "  -----\n  |   |\n  O   |\n      |\n      |\n=========",
  "  -----\n  |   |\n  O   |\n  |   |\n      |\n=========",
  "  -----\n  |   |\n  O   |\n /|   |\n      |\n=========",
  "  -----\n  |   |\n  O   |\n /|\\  |\n      |\n=========",
  "  -----\n  |   |\n  O   |\n /|\\  |\n /    |\n=========",
  "  -----\n  |   |\n  O   |\n /|\\  |\n / \\  |\n========="
]

def jugar
  palabra     = PALABRAS.sample
  adivinadas  = []
  intentos    = 0
  max         = DIBUJOS.length - 1

  loop do
    visible = palabra.chars.map { |l| adivinadas.include?(l) ? l : '_' }.join(' ')
    puts "\n#{DIBUJOS[intentos]}\n#{visible}\nIntentados: #{adivinadas.sort.join(', ')}"

    return puts('¡Ganaste!') unless visible.include?('_')
    return puts("¡Perdiste! La palabra era: #{palabra}") if intentos >= max

    print 'Letra: '
    letra = gets.chomp.downcase
    next puts('Ingresá una sola letra.') unless letra.length == 1 && letra.match?(/[a-z]/)
    next puts('Ya la intentaste.') if adivinadas.include?(letra)

    adivinadas << letra
    intentos += 1 unless palabra.include?(letra)
  end
end

puts '=== Ahorcado ==='
loop do
  jugar
  print '¿Jugar de nuevo? (s/n): '
  break unless gets.chomp.downcase == 's'
end

File: PHP/Basico/juego_ahorcado/ahorcado.php
<?php

$palabras = ['php', 'programacion', 'variable', 'funcion', 'array', 'clase', 'objeto', 'interfaz', 'bucle', 'puntero'];

$dibujos = [
    "  -----\n  |   |\n      |\n      |\n      |\n=========",
    "  -----\n  |   |\n  O   |\n      |\n      |\n=========",
    "  -----\n  |   |\n  O   |\n  |   |\n      |\n=========",
    "  -----\n  |   |\n  O   |\n /|   |\n      |\n=========",
    "  -----\n  |   |\n  O   |\n /|\\  |\n      |\n=========",
    "  -----\n  |   |\n  O   |\n /|\\  |\n /    |\n=========",
    "  -----\n  |   |\n  O   |\n /|\\  |\n / \\  |\n=========",
];

function jugar(array $palabras, array $dibujos): void
{
    $palabra    = $palabras[array_rand($palabras)];
    $adivinadas = [];
    $intentos   = 0;
    $max        = count($dibujos) - 1;

    while ($intentos < $max) {
        $visible = implode(' ', array_map(fn($l) => in_array($l, $adivinadas) ? $l : '_', str_split($palabra)));
        echo "\n{$dibujos[$intentos]}\n$visible\nIntentadas: " . implode(', ', $adivinadas) . "\n";

        if (!str_contains($visible, '_')) { echo "¡Ganaste!\n"; return; }

        echo 'Letra: ';
        $letra = strtolower(trim(fgets(STDIN)));
        if (strlen($letra) !== 1 || !ctype_alpha($letra)) { echo "Una sola letra.\n"; continue; }
        if (in_array($letra, $adivinadas)) { echo "Ya intentada.\n"; continue; }

        $adivinadas[] = $letra;
        if (!str_contains($palabra, $letra)) $intentos++;
    }
    echo "\n{$dibujos[$max]}\n¡Perdiste! La palabra era: $palabra\n";
}

echo "=== Ahorcado ===\n";
do {
    jugar($palabras, $dibujos);
    echo '¿Jugar de nuevo? (s/n): ';
} while (trim(fgets(STDIN)) === 's');
